File: AoC/2015/07.py
"""
"""

test = """
123 -> x
456 -> y
x AND y -> d
x OR y -> e
x LSHIFT 2 -> f
y RSHIFT 2 -> g
NOT x -> h
NOT y -> i
"""

from time import time

def TrackTime(func):
    def wrapper(*a, **k):
        start = time()
        func(*a, **k)
        print("  >> in %f ms" % ((time() - start)*1000))
    return wrapper

def ParseInput(filename):
    data = []
    with open(filename, 'r') as f:
        for l in f:
            data.append(l.strip())
    f.close()
    return data

def Affectation(i, values):
    values[i[2]] = int(i[0])

def And(i, values):
    values[i[4]] = values[i[0]] & values[i[2]]

def Or(i, values):
    values[i[4]] = values[i[0]] | values[i[2]]

@TrackTime
def Part1(data):
    total = 0
    values = {}
    for instruction in data:
        if not instruction: continue
        print(instruction)
        if 'AND' in instruction:
            And(instruction.split(), values)
        elif 'OR' in instruction:
            Or(instruction.split(), values)
        elif 'RSHIFT' in instruction:
            i = instruction.split()
            values[i[4]] = values[i[0]] >> int(i[2])
        elif 'LSHIFT' in instruction:
            i = instruction.split()
            values[i[4]] = values[i[0]] << int(i[2])
        elif 'NOT' in instruction:
            i = instruction.split()
            values[i[3]] = 65535 - int(values[i[1]])
        else:
            Affectation(instruction.split(), values)
        print(instruction, values)

    print(f'Answer to part 1 is {values['a']}')

@TrackTime
def Part2(data):
    total = 0
    print(f'Answer to part 1 is {total}')


data = ParseInput('./data/07.input')
# data = test.split('\n')
# print(data)
Part1(data)
Part2(data)

File: 2022/07.js
/**
 * Share your thoughts
 */

const year = '2022';
const day = '07';
const fs = require('fs');

const parseInput = () => {
  let filename = `./data/${day}`;
  if (process.argv.length > 2) {
    filename += '.test';
  }
  let data = fs.readFileSync(filename, 'utf8');
  return data;
}

const parseLines = (data) => {
  const lines = [];
  for (const line of data.split('\n')) {
    lines.push(line);
  }
  return lines;
}

const walkDirectories = (data) => {
  let summary = {};
  let currentDirectory = [];
  for (const line of data) {
    if (line.startsWith('$ cd')) {
      if (line === '$ cd ..') {
        currentDirectory.pop();
      } else {
        currentDirectory.push(currentDirectory.join('') + line.split(' ')[2]);
      }
    } else
    if (/^\d/.test(line)) {
      const size = parseInt(line.split(' ')[0]);
      for (const dir of currentDirectory) {
        if (dir in summary) {
          summary[dir] += size;
        } else {
          summary[dir] = size;
        }
      }
    }
  }
  return summary;
}

const part1 = (data) => {
  console.time('Time');
  let res = 0;

  const summary = walkDirectories(parseLines(data));
  for (const [key, size] of Object.entries(summary)) {
    if (size <= 100000) {
      res += size;
    }
  }

  console.log('Part 1:', res);
  console.timeEnd('Time');
}

const part2 = (data) => {
  console.time('Time');
  let res = 0;

  const max = 70000000;
  const target = 30000000;
  const summary = walkDirectories(parseLines(data));
  const goal = target - max + summary['/'];
  for (const [key, size] of Object.entries(summary)) {
    if (size > goal) {
      if (res === 0 || res > size) {
        res = size;
      }
    }
  }

  console.log('Part 2:', res);
  console.timeEnd('Time');
}

/**
 * Main
 */
console.log( `Advent Of Code ${year}-${day}`);
console.log('==========================');
part1(parseInput());
console.log('--------------------------');
part2(parseInput());
